### DISCUSSION TITLE
SQLite file locked by session in sqlalchemy 2.X

### DISCUSSION THREAD

**Discussion by pmithrandir** [sqlalchemy/sqlalchemy#11897] [Usage Questions] [answered]
Hello,

We are migrating from sqlalchemy 1.4 to 2.
Doing so, I think I found something fishy. 

Expectation: 
When ending a session, I expect to be able to remove my sqlite file as I was able to do in V1.4.

Code example:
```py
import os

from sqlalchemy import create_engine, Column, Integer, String, ForeignKey
from sqlalchemy.orm import sessionmaker, declarative_base, relationship

# cleaning before test
filename = "db_file.db"
if os.path.exists(filename):
    os.remove(filename)
# init DB
dsn = f"sqlite:///{filename}?check_same_thread=false"
engine = create_engine(dsn)

Session = sessionmaker(
    autocommit=False,
    autoflush=False,
    bind=engine)

Base = declarative_base()


class Human(Base):
    __tablename__ = "human"
    # __table_args__ = (UniqueConstraint('name', 'family_name', name="unique_person_name"),)

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False)

    house_id = Column(Integer, ForeignKey("house.id"), nullable=False)
    house = relationship(
        "House",
        back_populates="list_humans",
        foreign_keys=[house_id]
    )


class House(Base):
    __tablename__ = "house"
    id = Column(Integer, primary_key=True, index=True)
    value = Column(String(100), nullable=False)
    list_humans = relationship("Human", back_populates="house",
                               foreign_keys='[Human.house_id]')


Base.metadata.create_all(bind=engine)

# Creating objects
house = House(value="main_house")
human_pierre = Human(name="Pierre", house=house)
human_mathieu = Human(name="Mathieu", house=house)

# inject objects
with Session() as db_session:
    db_session.add(house)
    db_session.add(human_pierre)
    db_session.add(human_mathieu)
    db_session.commit()

# Trying to remove sqlite file <----BUG is HERE
os.remove(filename)

```

This code works great on 1.4.
On 2.0.35 it doesn't and raise the exception:
 ```
Traceback (most recent call last):
  File "C:\Users\XXXXXX\PycharmProjects\XXXXX\scripts\sandbox\sqlalchemy_bug.py", line 60, in <module>
    os.remove(filename)
PermissionError: [WinError 32] Le processus ne peut pas accéder au fichier car ce fichier est utilisé par un autre processus: 'db_file.db'
```
which means the process can't access the file because it's used by another process.

I know I can use dispose() function to release the connection, but this example is drastically simplier than our entire application, and it still doesn't seem right to keep connection active like that when session is finished.

Also, in my real project, If I wait around 10 sec before removing the file, it is sometimes working. 

Discussion https://github.com/sqlalchemy/sqlalchemy/discussions/11709 seems close, but they actually avoided the issue more than resolving it.

Best regards,
Pierre

**Comment by pmithrandir** [comment]
BTW, I tried to use NullPool.
It works on that script, but affect some other part of my code based on in_memory sqlite db which seems to be incompatible with NullPool.

**Reply by gordthompson** [reply, marked as answer]
> it still doesn't seem right to keep connection active like that when session is finished

In 1.4, the default for a file-based SQLite database was `NullPool`. In 2.0 the default is now `QueuePool`, and that's how `QueuePool` works.

https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-aba109875059299e0eb82359b5825416

> other part of my code based on in_memory sqlite db which seems to be incompatible with NullPool

A different database (in-memory vs. file-based) suggests using a separate engine, which would use a separate pool.

**Reply by zzzeek** [reply]
well nullpool is incompatible with a memory database, as is QueuePool.  if you use a memory database it's going to pick SingletonThreadPool.   There's no reason a program would have any cross-confusion here though, if the code does something with SQLite file databases, then it should do that only for the file DBs and not for any memory DBs

**Comment by pmithrandir** [comment]
Thank you both for your  answers.

Our software is a sort of custom ETL.
We Connect to many external DB, store a "view" of that DB converted to our internal model.
We use sqlite to store db export on file, or to execute internal treatment using sqlite in memory.

But really, the time we use the most sqlite is for our tests. We execute the same code that we use for our DBs, but on sqlite, if possible in_memory for performance reason.

I'll try tomorrow to make sure sqlite DB are using the SingletonThreadPool. It will be easy to find out if it's working.

Outside of that, I think I spent the last 5 days working full time on the sqlalchemy migration and I don't recall seeing alert / information in the migration guide regarding the change of Pool selected by default. As they might be incompatible (as in my example) it might be interesting to make it more clear.

Thanks again !

Pierre

**Reply by gordthompson** [reply]
> information in the migration guide regarding the change of Pool selected by default

https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#the-sqlite-dialect-uses-queuepool-for-file-based-databases

**Reply by pmithrandir** [reply]
thank you, didn't catch it. (I was using the step by step one I believe)

**Comment by pmithrandir** [comment]
Hello,

Thank you both for your help.
I was able to make it work again using NullPool for all my sqllite DB using files. (in memory are still with default value)

Pierre
